compiler/resolution/resolveFunction.cpp

```diff
@@ -984,6 +984,109 @@ static void computeReturnTypeParamVectors(BaseAST*      ast,
                                           Vec<Type*>&   retTypes,
                                           Vec<Symbol*>& retSymbols);
 
+void resolveIfExprType(CondStmt* stmt) {
+  Vec<Type*> retTypes;
+  Vec<Symbol*> retSymbols;
+
+  CallExpr* last = toCallExpr(stmt->thenStmt->body.tail);
+  INT_ASSERT(last->isPrimitive(PRIM_MOVE));
+  Symbol* ret = toSymExpr(last->get(1))->symbol();
+  FnSymbol* fn = toFnSymbol(stmt->parentSymbol); // dummy to use for canDispatch
+
+  computeReturnTypeParamVectors(stmt, ret, retTypes, retSymbols);
+
+  if (retTypes.n == 1) {
+    ret->type = retTypes.head();
+  } else {
+    INT_ASSERT(retTypes.n == 2 && retSymbols.n == 2);
+
+    // NB: Assumes '0' and '1' correspond to 'then' and 'else', respectively.
+    Type* thenType = retTypes.v[0];
+    Type* elseType = retTypes.v[1];
+    Type* retType  = NULL;
+
+    Symbol* thenSym = retSymbols.v[0];
+    Symbol* elseSym = retSymbols.v[1];
+
+    bool thenTypeVar = thenSym->hasFlag(FLAG_TYPE_VARIABLE);
+    bool elseTypeVar = elseSym->hasFlag(FLAG_TYPE_VARIABLE);
+
+    // If only one branch returns a reference, that branch needs to copy the
+    // result so that we can always free the result.
+    if (isReferenceType(thenType) != isReferenceType(elseType)) {
```

Comment:
Is there a reason this only works with reference types and not also QUAL_REF? I thought we had another function (like isReference maybe?) that handled the Qualifier case as well?

Reply:
I just copied what ``resolveReturnTypeAndYieldedType`` did, and it uses ``isReferenceType``. For some reason I have it in my head that there are issues with ``isRef`` during resolution.

Reply:
There are, but it's mostly that resolution doesn't use QUAL_REF. Anyway you can just add a comment about how it should eventually be updated to handle the qualifiers.